### PR TITLE
feat(rust, sql): sql udfs

### PR DESCRIPTION
**Pull request by universalmind303**
currently polars sql has no mechanism for defining custom functions. 

This adds a new `function_registry` to the sql context. Users can implement this trait to provide a registry of custom functions that will be used when parsing & executing the sql statements. 

I'm not a python expert, but it does look like we have significant functionality around python UDF's. Considering we are already converting python functions to SeriesUdf, it should be possible to expose this to python's sql context as well. 

see full example in `examples/sql_udfs/src/main.rs` for expected usage.

### CODE REVIEW

**Comment by lucazanna** [comment]
Hi @universalmind303 

if this was exposed to Python, would it be similar to an Expr.pipe (https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.pipe.html#polars.Expr.pipe) or to a map (so with a Python UDF)?

Nice development !

**Comment by universalmind303** [comment]
hey @ritchie46 can you take a look at this when you have a moment.

File: crates/polars-sql/src/function_registry.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    fn contains(&self, _name: &str) -> bool {
+        false
+    }
+}
+
+/// Represents a user-defined function that can be used within the system.
+///
+/// This trait provides several methods to describe and manipulate the user-defined function.
+/// This is to be used with the [`FunctionRegistry`].
+pub trait UserDefinedFunction: Send + Sync {
```
Comment (ritchie46) [inline crates/polars-sql/src/function_registry.rs:38]: Now we have 2 traits that define a UDF. Can we rewrite this to use `SeriesUdf`. We should consolidate functionality.

File: crates/polars-sql/src/context.rs [hunk 2]
```diff
@@ -107,6 +115,23 @@ impl SQLContext {
         self.cte_map.borrow_mut().clear();
         res
     }
+
+    /// add a function registry to the SQLContext
+    /// the registry provides the ability to add custom functions to the SQLContext
+    pub fn with_function_registry(mut self, function_registry: Arc<dyn FunctionRegistry>) -> Self {
+        self.function_registry = function_registry;
+        self
```
Comment (SeanTroyUWO) [inline crates/polars-sql/src/context.rs:123]: If I was using this, I'd assume that the registries would combine, not overwrite. When you do `with_context`, that's how it works.

File: examples/sql_udfs/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        },
+    );
+
+    // register a new UDF on an existing context
+    ctx.registry_mut().register("my_div", my_custom_divide)?;
+
+    // execute the query
+    let res = ctx.execute("SELECT a, b, my_div(a, b) as my_div FROM foo")?;
+
+    println!("{:?}", res.collect()?);
```
Comment (SeanTroyUWO) [inline examples/sql_udfs/src/main.rs:91]: Would it be possible to toss this into a test?

File: crates/polars-sql/src/context.rs (hunk 2 above)
Comment (universalmind303) [reply crates/polars-sql/src/context.rs:123]: a lot of the other structs such as the readers/writers use this pattern & is a pretty common alternative to the builder pattern. 
- [csv reader](https://github.com/universalmind303/polars/blob/410f869de8356df381d8ee3c7cbc38186dd76e04/crates/polars-io/src/csv/read.rs#L139)
- [csv writer](https://github.com/universalmind303/polars/blob/410f869de8356df381d8ee3c7cbc38186dd76e04/crates/polars-io/src/csv/write.rs#L60)
- [cloud options](https://github.com/universalmind303/polars/blob/410f869de8356df381d8ee3c7cbc38186dd76e04/crates/polars-core/src/cloud.rs#L94)

File: examples/sql_udfs/src/main.rs (hunk 3 above)
Comment (universalmind303) [reply examples/sql_udfs/src/main.rs]: moved this to a test in `crates/polars-sql/tests/udf.rs`

**Review by ritchie46 — APPROVED** [review]
Looks good @universalmind303. I still feel that we create some double functionality with `UserDefinedFunction` and `SeriesUdf`. maybe we should consolidate both into a single thing.

Can you make CI happy so we can merge?

**Pull request merged**